src/drawer.tsx

```diff
@@ -0,0 +1,306 @@
+ */
+const MaybeDivider = ({show}: {show: boolean}) =>
+    show ? <Divider /> : <div style={{borderBottom: '1px solid transparent'}} />;
+
+/**
+ * Restores the focus to the element that was focused before the Drawer was opened
+ */
+const useRestoreFocus = () => {
+    const activeElementRef = React.useRef<HTMLElement | null>(document.activeElement as HTMLElement);
+    React.useEffect(() => {
+        const elementToFocus = activeElementRef.current;
+        return () => {
+            elementToFocus?.focus?.();
+        };
+    }, []);
+};
+
+type DrawerLayoutProps = {
+    width?: number;
+    children: React.ReactNode;
+    onClose: () => void;
+    onDismiss?: () => void;
+};
+
+type DrawerPropsRef = {
+    close: () => Promise<void>;
+    dismiss: () => Promise<void>;
+};
+
+const DrawerLayout = React.forwardRef<DrawerPropsRef, DrawerLayoutProps>(
+    ({width, children, onClose, onDismiss}, ref) => {
+        useSetModalStateEffect();
+        useRestoreFocus();
+        const {isMobile, isTablet} = useScreenSize();
+        const [isOpen, setIsOpen] = React.useState(false);
+        const widthStyle = isMobile ? 'auto' : width || (isTablet ? WIDTH_TABLET : WIDTH_DESKTOP);
+
+        const open = React.useCallback((node: HTMLDivElement) => {
+            if (node) {
+                // small delay to allow the Portal to be mounted
+                setTimeout(() => {
+                    setIsOpen(true);
+                }, 50);
+            }
+        }, []);
+
+        const close = React.useCallback(() => {
+            setIsOpen(false);
+            return new Promise((resolve) => {
+                setTimeout(resolve, styles.ANIMATION_DURATION_MS);
+            }).then(onClose);
+        }, [onClose]);
+
+        const dismiss = React.useCallback(() => {
+            return close().then(() => onDismiss?.());
+        }, [onDismiss, close]);
+
+        React.useImperativeHandle(ref, () => ({close, dismiss}));
+
+        React.useEffect(() => {
+            const handleKeyDown = (event: KeyboardEvent) => {
+                if (event.key === 'Escape') {
+                    dismiss();
+                }
+            };
+            document.addEventListener('keydown', handleKeyDown);
+            return () => {
+                document.removeEventListener('keydown', handleKeyDown);
+            };
+        }, [dismiss]);
+
+        return (
+            <Portal>
+                <FocusTrap>
+                    {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
+                    <div
+                        data-testid="drawerOverlay"
+                        onClick={onDismiss ? dismiss : undefined}
+                        className={classnames(
+                            styles.overlay,
+                            isOpen ? styles.overlayOpen : styles.overlayClosed
+                        )}
+                    />
+                    <div
+                        data-testid="drawerLayout"
```

Comment:
is this testid needed?

Reply:
Not really. I've added it to differentiate the boundaries of the different parts of this component. The `DrawerLayout` is a separate component as `Drawer`.